pyro/infer/trace_elbo.py

```diff
@@ -9,6 +8,21 @@
 from pyro.infer.elbo import ELBO
 from pyro.poutine.util import prune_subsample_sites
 from pyro.util import check_model_guide_match, check_site_shape, is_nan
+from pyro.infer.util import MultiFrameTensor
+
+
+# TODO cache some of these computations
+def compute_site_log_r(model_trace, guide_trace, target_site):
```

Comment:
With a tiny bit of refactoring you could reduce from quadratic to linear time. First return `log_r` in this function:
```py
def compute_log_r(model_trace, guide_trace):
    ...your existing code...
    return log_r
```
Then lazily compute and cache `log_r` in `Trace_ELBO.loss_and_grads()`:
```py
log_r = None  # compute lazily
for name, model_site in model_trace.nodes.items():
    if model_site["type"] == "sample":
        ...
    else:
        ...
        if not is_identically_zero(score_function_term):
            if log_r is None:
                log_r = compute_log_r(model_trace, guide_trace)
            log_r_site = log_r.sum_t(guide_site["cond_indep_stack"])
            ...
```